compiler/resolution/functionResolution.cpp

```diff
@@ -927,22 +954,73 @@ determineQueriedField(CallExpr* call) {
 }
 
 
-//
-// For some types, e.g. _domain/_array records, implementing
-// Chapel's ref/out/... intents can be done simply by passing the
-// value itself, rather than address-of. This function flags such cases
-// by returning false, meaning "not OK to convert".
-//
+// Generally speaking, tuples containing refs should be converted
+// to tuples without refs before returning.
+// This function returns true for exceptional FnSymbols
+// where tuples containing refs can be returned.
 static bool
-okToConvertFormalToRefType(Type* type) {
-  if (isRecordWrappedType(type))
-    // no, don't
+doNotChangeTupleTypeRefLevel(FnSymbol* fn, bool forRet)
+{
+
+  if( fn->hasFlag(FLAG_TYPE_CONSTRUCTOR) || // but not _type_construct__tuple
+      fn->hasFlag(FLAG_CONSTRUCTOR) || // but not _construct__tuple
+      fn->hasFlag(FLAG_BUILD_TUPLE) || // and not _build_tuple(_allow_ref)
+      fn->hasFlag(FLAG_BUILD_TUPLE_TYPE) || // and not _build_tuple_type
+      fn->hasFlag(FLAG_TUPLE_CAST_FN) || // and not _cast for tuples
+      fn->hasFlag(FLAG_EXPAND_TUPLES_WITH_VALUES) || // not iteratorIndex
+      (!forRet && fn->hasFlag(FLAG_INIT_COPY_FN)) || // not tuple chpl__initCopy
+      fn->hasFlag(FLAG_AUTO_COPY_FN) || // not tuple chpl__autoCopy
+      fn->hasFlag(FLAG_AUTO_DESTROY_FN) || // not tuple chpl__autoDestroy
+      fn->hasFlag(FLAG_UNALIAS_FN) || // not tuple chpl__unalias
+      fn->hasFlag(FLAG_ALLOW_REF) || // not iteratorIndex
+      (forRet && fn->hasFlag(FLAG_ITERATOR_FN)) // iterators b/c
+                                    //  * they might return by ref
+                                    //  * might need to return a ref even
+                                    //    when not indicated return by ref.
+      //(fn->hasFlag(FLAG_ITERATOR_FN) && // iterators returning by ref
+      // (fn->retTag == RET_REF ||        // might yield index variable by
+      //  fn->retTag == RET_CONST_REF))   // ref in a tuple
+     ) {
+    return true;
+  } else {
```

Comment:
I removed the last three, at least. This whole function is basically one big work-around so I think it's useful to add any context at all. Some of the comments aro enot obvious (e.g. that FLAG_ALLOW_REF is there to disqualify iteratorIndex).